### PR TITLE
[#198] feature/external attendee booking

### PR DESCRIPTION
**Pull request by Pavliuk13**
https://user-images.githubusercontent.com/59260347/188367466-2bf66627-f639-4e10-95e0-cf59a0542154.mp4

### CODE REVIEW

File: frontend/src/app/modules/availability/availability-page/availability-page.component.ts [hunk 1]
```diff
@@ -49,7 +49,7 @@ export class AvailabilityPageComponent extends BaseComponent {
                 if (resp) {
                     this.currentUser = resp;
                     this.getUserPersonalAndTeamSlots();
-                    this.teamService.currentTeamEmitted$.subscribe(teamId => {
+                    this.teamService.currentTeamEmitted$.subscribe((teamId) => {
```
Comment (EugeneValko) [inline frontend/src/app/modules/availability/availability-page/availability-page.component.ts]: You need to unsubscripe

File: frontend/src/app/modules/bookings/bookings-page/bookings-page.component.ts [hunk 2]
```diff
@@ -25,10 +25,9 @@ export class BookingsPageComponent extends BaseComponent implements OnInit {
     public meetings: IMeetingBooking[];
 
     public ngOnInit(): void {
-        this.teamService.currentTeamEmitted$
-            .subscribe(teamId => {
-                this.loadMeetings(teamId);
-            });
+        this.teamService.currentTeamEmitted$.subscribe((teamId) => {
+            this.loadMeetings(teamId);
+        });
```
Comment (EugeneValko) [inline frontend/src/app/modules/bookings/bookings-page/bookings-page.component.ts]: You need to unsubscripe

File: frontend/src/app/modules/external-booking/external-booking-choose-meeting-page/external-booking-choose-meeting-page.component.html [hunk 3]
```diff
@@ -1,19 +1,15 @@
...
-            [queryParams]="{ duration: item.size }"
-            (click)="addDurationAndLocation(item.size, item.locationType)">
-            <mat-card-title><p class="title">{{ item.name }}</p></mat-card-title>
-            <mat-card-subtitle>
-                {{ item.size }} Min, {{ locationTypeMapping(item.locationType)}}
-            </mat-card-subtitle>
+        <mat-card class="card" [routerLink]="['/external-booking/choose-time/']" [queryParams]="{ link: item.link }">
+            <mat-card-title
+                ><p class="title">{{ item.name }}</p></mat-card-title
+            >
```
Comment (EugeneValko) [inline frontend/src/app/modules/external-booking/external-booking-choose-meeting-page/external-booking-choose-meeting-page.component.html]:
```suggestion
            <mat-card-title><p class="title">{{ item.name }}</p></mat-card-title>
```

File: frontend/src/app/modules/external-booking/external-booking-choose-time-page/external-booking-choose-time-page.component.ts [hunk 4]
```diff
@@ -42,35 +59,50 @@ export class ExternalBookingTimeComponent extends BaseComponent implements OnIni
...
-                this.selectedUserAvailabilitySlots = slots;
-            });
     }
 
     ngOnInit(): void {
         this.calendarWeek = this.getCurrentWeek();
         this.route.queryParams.subscribe((params) => {
-            this.selectedMeetingDuration = params['duration'];
+            this.link = params['link'];
         });
```
Comment (EugeneValko) [inline frontend/src/app/modules/external-booking/external-booking-choose-time-page/external-booking-choose-time-page.component.ts]: You need to unsubscripe

File: frontend/src/app/modules/external-booking/external-booking-confirm-page/external-booking-confirm-page.component.html [hunk 5]
```diff
@@ -1,25 +1,40 @@
...
+        <input class="input-message" [(ngModel)]="externalAnswers.externalName" formControlName="characters" />
+        <span *ngIf="confirmForm.controls['characters'].hasError('minlength')"
+            ><p class="validationError">The minimum length is 2, the maximum is 50.</p></span
+        >
+        <span *ngIf="confirmForm.controls['characters'].hasError('pattern')"
+            ><p class="validationError">
+                Invalid input field. Allow only only text, numbers, the symbol -. <br />Space cannot be the first or
+                last character.
+            </p></span
+        >
```
Comment (EugeneValko) [inline frontend/src/app/modules/external-booking/external-booking-confirm-page/external-booking-confirm-page.component.html]:
```suggestion
        <span *ngIf="confirmForm.controls['characters'].hasError('minlength')">
                <p class="validationError">The minimum length is 2, the maximum is 50.</p>
        </span>
        <span *ngIf="confirmForm.controls['characters'].hasError('pattern')">
                <p class="validationError">
                Invalid input field. Allow only only text, numbers, the symbol -. <br />Space cannot be the first or
                last character.</p>
        </span>
```

File: frontend/src/app/modules/external-booking/external-booking-confirm-page/external-booking-confirm-page.component.html [hunk 6]
```diff
@@ -1,25 +1,40 @@
...
     </div>
     <div id="container-email" class="email-container" [formGroup]="confirmForm">
         <p class="input-title">Your Email</p>
-        <textarea class="input-message" formControlName="email" [(ngModel)]="externalAnswers.externalEmail"></textarea>
-        <span *ngIf="confirmForm.controls['email'].hasError('email')"><p class="validationError">Email format is invalid.</p></span>
+        <input class="input-message" [(ngModel)]="externalAnswers.externalEmail" formControlName="email" />
+        <!-- <textarea class="input-message" formControlName="email" [(ngModel)]="externalAnswers.externalEmail"></textarea> -->
+        <span *ngIf="confirmForm.controls['email'].hasError('email')"
+            ><p class="validationError">Email format is invalid.</p></span
+        >
```
Comment (EugeneValko) [inline frontend/src/app/modules/external-booking/external-booking-confirm-page/external-booking-confirm-page.component.html]:
```suggestion
        <span *ngIf="confirmForm.controls['email'].hasError('email')"
            ><p class="validationError">Email format is invalid.</p></span>
```

File: frontend/src/app/modules/external-booking/external-booking-confirm-page/external-booking-confirm-page.component.html [hunk 7]
```diff
@@ -1,25 +1,40 @@
...
-        <textarea class="input-message" formControlName="extraInfo" [(ngModel)]="externalAnswers.externalExtraInfo"></textarea>
-        <span *ngIf="confirmForm.controls['extraInfo'].hasError('minlength')"><p class="validationError">The minimum length is 3, the maximum is 80.</p></span>
+        <p class="input-title">Please share anything that will help prepare<br />for our meeting</p>
+        <textarea
+            class="additional-info"
+            formControlName="extraInfo"
+            [(ngModel)]="externalAnswers.externalExtraInfo"></textarea>
+        <span *ngIf="confirmForm.controls['extraInfo'].hasError('minlength')"
+            ><p class="validationError">The minimum length is 3, the maximum is 80.</p></span
+        >
```
Comment (EugeneValko) [inline frontend/src/app/modules/external-booking/external-booking-confirm-page/external-booking-confirm-page.component.html]:
```suggestion
        <span *ngIf="confirmForm.controls['extraInfo'].hasError('minlength')"
            ><p class="validationError">The minimum length is 3, the maximum is 80.</p></span>
```

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/MappingProfiles/MeetingProfile.cs [hunk 8]
```diff
@@ -12,7 +12,7 @@ public MeetingProfile()
             CreateMap<User, UserMeetingDTO>()
                 .ForMember(d => d.TimeZone, s => s.MapFrom(s => s.TimeZone.ToString()));
             CreateMap<ExternalAttendee, UserMeetingDTO>()
-                .ForMember(d => d.TimeZone, s => s.MapFrom(s => s.TimeZone.ToString()));
+                .ForMember(d => d.TimeZone, s => s.MapFrom(s => s.TimeZone.ToString()));;
             CreateMap<Meeting, MeetingThreeMembersDTO>()
```
Comment (EugeneValko) [inline backend/EasyMeets.Core/EasyMeets.Core.BLL/MappingProfiles/MeetingProfile.cs]: why change the TimeZone in string?

**Review by EugeneValko — APPROVED** [review]

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/ExternalAttendeeService.cs [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+    public ExternalAttendeeService(EasyMeetsCoreContext context, IMapper mapper) : base(context, mapper) { }
+
+    public async Task SaveExternalAttendeeAsync(ExternalAttendeeAndBookedMeetingDto bookingDto)
+    {
+        var meeting = _mapper.Map<Meeting>(bookingDto.Meeting);
+        var createdMeeting = _context.Meetings.Add(meeting).Entity;
+        await _context.SaveChangesAsync();
+        
+        var attendee = _mapper.Map<ExternalAttendee>(bookingDto.Attendee);
+        attendee.MeetingId = createdMeeting.Id;
```
Comment (apofis96) [inline backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/ExternalAttendeeService.cs:20]: Doesn't meeting var contain id after saveChanges()?

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/MeetingService.cs [hunk 10]
```diff
@@ -96,6 +97,15 @@ public async Task CreateMeeting(SaveMeetingDto meetingDto)
...
         }
 
+        public async Task<List<OrderedMeetingTimesDto>> GetOrderedMeetingTimesAsync(long slotId)
+        {
+            var result = await _context.Meetings.Where(el => el.AvailabilitySlotId == slotId)
+                .Select(el => new OrderedMeetingTimesDto {StartTime = el.StartTime})
+                .ToListAsync();
+
+            return result;
+        }
```
Comment (apofis96) [inline backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/MeetingService.cs]:
```suggestion
        public Task<List<OrderedMeetingTimesDto>> GetOrderedMeetingTimesAsync(long slotId)
        {
            return  _context.Meetings.Where(el => el.AvailabilitySlotId == slotId)
                .Select(el => new OrderedMeetingTimesDto {StartTime = el.StartTime})
                .ToListAsync();
        }
```

File: frontend/src/app/modules/external-booking/external-booking-choose-time-page/external-booking-choose-time-page.component.ts [hunk 11]
```diff
@@ -82,64 +114,104 @@ export class ExternalBookingTimeComponent extends BaseComponent implements OnIni
...
+        let isOrdered = false;
+
+        const firstCalendarDay = new Date(this.calendarWeek.firstDay);
+
+        firstCalendarDay.setHours(date.getHours(), date.getMinutes(), date.getSeconds(), date.getMilliseconds());
+        const result = addDays(firstCalendarDay, daysToAdd);
+
+        result.setTime(result.getTime() + this.selectedMeetingDuration * timesToAdd * 60 * 1000);
+
+        this.orderedTimes.forEach((el) => {
```
Comment (apofis96) [inline frontend/src/app/modules/external-booking/external-booking-choose-time-page/external-booking-choose-time-page.component.ts]: Cannot findIndex (or some, if you wish) be used here?

**Review by apofis96 — CHANGES_REQUESTED** [review]

File: backend/EasyMeets.Core/EasyMeets.Core.DAL/Migrations/20220902032122_ChangeAvailabilitySetting.cs [hunk 12]
```diff
@@ -151,7 +151,7 @@ protected override void Up(MigrationBuilder migrationBuilder)
                 keyColumn: "Id",
                 keyValue: 7L,
                 columns: new[] { "Color", "UpdatedAt" },
-                values: new object[] { 8, new DateTime(2022, 9, 2, 0, 0, 0, 0, DateTimeKind.Local) });
+                values: new object[] { 7, new DateTime(2022, 9, 2, 0, 0, 0, 0, DateTimeKind.Local) });
```
Comment (tatianahutii) [inline backend/EasyMeets.Core/EasyMeets.Core.DAL/Migrations/20220902032122_ChangeAvailabilitySetting.cs]: Why did you change existing migration?

File: frontend/src/app/core/helpers/schedule-items-helper.ts [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+        const newItem: IScheduleItemReceive = {
+            ...el,
+            start: newStart,
+            end: newEnd,
+        };
+
+        return newItem;
+    });
+
+    return newArray;
```
Comment (tatianahutii) [inline frontend/src/app/core/helpers/schedule-items-helper.ts]: you can combine these lines

File: frontend/src/app/modules/external-booking/external-booking-choose-time-page/external-booking-choose-time-page.component.ts [hunk 14]
```diff
@@ -82,64 +114,104 @@ export class ExternalBookingTimeComponent extends BaseComponent implements OnIni
...
-        const timeFinish = new Date(date.getTime() + this.selectedMeetingDuration * 60000);
+    public AddTimeAndDate(timeIndex: number, dayIndex: number, timeZone: TimeZone): void {
+        const date = addDays(this.calendarWeek.firstDay, dayIndex);
+        const time = addHours(this.theEarliestStartOfTimeRanges, (this.selectedMeetingDuration * timeIndex) / 60);
 
-        this.selectedTimeAndDateEvent.emit({ date, timeFinish });
+        date.setHours(time.getHours());
+        date.setMinutes(time.getMinutes());
+        date.setSeconds(time.getSeconds());
+        date.setMilliseconds(time.getMilliseconds());
```
Comment (tatianahutii) [inline frontend/src/app/modules/external-booking/external-booking-choose-time-page/external-booking-choose-time-page.component.ts]: I believe you can combine these lines

File: frontend/src/app/modules/external-booking/external-booking-choose-time-page/external-booking-choose-time-page.component.ts [hunk 15]
```diff
@@ -82,64 +114,104 @@ export class ExternalBookingTimeComponent extends BaseComponent implements OnIni
...
     public changeWeek(addingMode: boolean): void {
         if (addingMode) {
-            this.calendarWeek.firstDay = new Date(
-                this.calendarWeek.firstDay.setDate(this.calendarWeek.firstDay.getDate() + 7),
-            );
-            this.calendarWeek.lastDay = new Date(
-                this.calendarWeek.lastDay.setDate(this.calendarWeek.lastDay.getDate() + 7),
-            );
+            this.calendarWeek.firstDay = addDays(this.calendarWeek.firstDay, 7);
+            this.calendarWeek.lastDay = addDays(this.calendarWeek.lastDay, 7);
```
Comment (tatianahutii) [inline frontend/src/app/modules/external-booking/external-booking-choose-time-page/external-booking-choose-time-page.component.ts]: get rid of mutation

File: frontend/src/app/modules/external-booking/external-booking-choose-time-page/external-booking-choose-time-page.component.ts [hunk 16]
```diff
@@ -82,64 +114,104 @@ export class ExternalBookingTimeComponent extends BaseComponent implements OnIni
...
         } else {
-            this.calendarWeek.firstDay = new Date(
-                this.calendarWeek.firstDay.setDate(this.calendarWeek.firstDay.getDate() - 7),
-            );
-            this.calendarWeek.lastDay = new Date(
-                this.calendarWeek.lastDay.setDate(this.calendarWeek.lastDay.getDate() - 7),
-            );
+            this.calendarWeek.firstDay = subDays(this.calendarWeek.firstDay, 7);
+            this.calendarWeek.lastDay = subDays(this.calendarWeek.lastDay, 7);
```
Comment (tatianahutii) [inline frontend/src/app/modules/external-booking/external-booking-choose-time-page/external-booking-choose-time-page.component.ts]: see the previous comment

**Review by tatianahutii — CHANGES_REQUESTED** [review]

File: frontend/src/app/modules/external-booking/external-booking-choose-time-page/external-booking-choose-time-page.component.ts (hunk 11 above)
Comment (Pavliuk13) [reply frontend/src/app/modules/external-booking/external-booking-choose-time-page/external-booking-choose-time-page.component.ts]: We cannot use it ( We must check every element in this.orderedTimes because it could be any date in any order.

File: backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/ExternalAttendeeService.cs (hunk 9 above)
Comment (Pavliuk13) [reply backend/EasyMeets.Core/EasyMeets.Core.BLL/Services/ExternalAttendeeService.cs:20]: meeting variable doesn't contain id after saveChanges(). I checked this

File: frontend/src/app/modules/external-booking/external-booking-choose-time-page/external-booking-choose-time-page.component.ts [hunk 17]
```diff
@@ -80,64 +110,101 @@ export class ExternalBookingTimeComponent extends BaseComponent implements OnIni
...
+
+        result.setTime(result.getTime() + this.selectedMeetingDuration * timesToAdd * 60 * 1000);
+
+        for (let i = 0; i < this.orderedTimes.length; i++) {
+            const parsedDate = new Date(Date.parse(this.orderedTimes[i].startTime));
+
+            if (parsedDate.getDate() === result.getDate() && parsedDate.getTime() === result.getTime()) {
+                return false;
+            }
+        }
```
Comment (apofis96) [inline frontend/src/app/modules/external-booking/external-booking-choose-time-page/external-booking-choose-time-page.component.ts]:
```suggestion
        const Test = (parsedDate: Date) =>
            parsedDate.getDate() === result.getDate() && parsedDate.getTime() === result.getTime();

        const index = this.orderedTimes.findIndex(x => Test(new Date(Date.parse(x.startTime))));

        if (index !== -1) {
            return false;
        }
```

**Review by apofis96 — CHANGES_REQUESTED** [review]

**Review by tatianahutii — APPROVED** [review]

**Review by apofis96 — APPROVED** [review]
